cpp/path_module/algorithm/path.hpp

```diff
@@ -56,52 +57,102 @@ struct LabelBoolsStatus {
+  std::unordered_map<std::string, RelDirection> relationship_sets;
+  LabelSets label_sets;
+  int64_t min_hops = 0;
+  int64_t max_hops = std::numeric_limits<int64_t>::max();
+  bool any_incoming = false;
+  bool any_outgoing = false;
+  bool filter_start_node = true;
+  bool begin_sequence_at_start = true;
+  bool bfs = false;
 };
 
-void Create(mgp_list *args, mgp_graph *memgraph_graph, mgp_result *result, mgp_memory *memory);
+class PathHelper {
+ public:
+  explicit PathHelper(const mgp::List &labels, const mgp::List &relationships, int64_t min_hops, int64_t max_hops);
+  explicit PathHelper(const mgp::Map &config);
 
-void FilterLabelBoolStatus(const LabelSets &labelSets, LabelBoolsStatus &labelStatus);
+  RelDirection GetDirection(std::string &rel_type);
+  LabelBools GetLabelBools(const mgp::Node &node);
 
-bool ShouldExpand(const LabelBools &labelBools, const LabelBoolsStatus &labelStatus);
+  bool AnyDirected(bool outgoing) const { return outgoing ? config_.any_outgoing : config_.any_incoming; }
+  bool FilterNodes(bool is_start) const { return (config_.filter_start_node || !is_start); }
+  bool FilterRelationships(bool is_start) const { return (config_.begin_sequence_at_start || !is_start); }
 
-void Expand(mgp_list *args, mgp_graph *memgraph_graph, mgp_result *result, mgp_memory *memory);
+  bool AreLabelsValid(const LabelBools &label_bools) const;
+  bool ContinueExpanding(const LabelBools &label_bools, size_t path_size) const;
 
-void FilterLabel(const std::string_view label, const LabelSets &labelFilters, LabelBools &labelBools);
+  bool PathSizeOk(int64_t path_size) const;
+  bool PathTooBig(int64_t path_size) const;
+  bool Whitelisted(bool whitelisted) const;
 
-void ParseLabels(const mgp::List &list_of_labels, LabelSets &labelSets);
+  // methods for parsing config
+  void FilterLabelBoolStatus();
+  void FilterLabel(std::string_view label, LabelBools &label_bools);
+  void ParseLabels(const mgp::List &list_of_labels);
+  void ParseRelationships(const mgp::List &list_of_relationships);
 
-void ParseRelationships(const mgp::List &list_of_relationships, RelationshipSets &relationshipSets, bool &any_outgoing,
-                        bool &any_incoming);
-bool PathSizeOk(const int64_t path_size, const int64_t &max_hops, const int64_t &min_hops);
+ private:
+  Config config_;
+};
 
-bool RelationshipAllowed(const std::string &rel_type, const RelationshipSets &relationshipSets, bool &any_outgoing,
-                         bool &any_incoming, bool outgoing);
+class PathData {
+ public:
+  explicit PathData(const PathHelper &helper, const mgp::RecordFactory &record_factory, const mgp::Graph &graph)
+      : helper_(helper), record_factory_(record_factory), graph_(graph) {}
+
+ protected:
+  PathHelper helper_;
+  const mgp::RecordFactory &record_factory_;
+  const mgp::Graph &graph_;
+  std::unordered_set<int64_t> visited_;
+};
 
-bool Whitelisted(const bool &whitelisted, const bool &whitelist_empty);
+class PathExpand : PathData {
```

Comment:
I did it since PathSubgraph and PathExpand share a lot of variables so they be just called with variable instead of class.variable, but I guess that's not a good practice

Reply:
I think its the simplest to just add PathData variables to both classes since otherwise it would be something like path_data_.getHelper().something(). And it's only 4 shared variables.